dart/simulation/World.cpp

```diff
@@ -74,11 +76,49 @@ World::~World()
   delete mConstraintSolver;
   delete mRecording;
 
-  for (std::vector<dynamics::Skeleton*>::const_iterator it = mSkeletons.begin();
-       it != mSkeletons.end(); ++it)
+  for(common::Connection& connection : mNameConnectionsForSkeletons)
+    connection.disconnect();
+
+  for(common::Connection& connection : mNameConnectionsForFrames)
+    connection.disconnect();
+}
+
+//==============================================================================
+WorldPtr World::clone() const
+{
+  WorldPtr worldClone(new World(mName));
+
+  worldClone->setGravity(mGravity);
+  worldClone->setTimeStep(mTimeStep);
+
+  // Clone and add each Skeleton
+  for(size_t i=0; i<mSkeletons.size(); ++i)
+  {
+    worldClone->addSkeleton(mSkeletons[i]->clone());
+  }
+
+  // Clone and add each SimpleFrame
+  for(size_t i=0; i<mFrames.size(); ++i)
+  {
+    worldClone->addFrame(mFrames[i]->clone(mFrames[i]->getParentFrame()));
+  }
+
+  // For each newly cloned SimpleFrame, try to make its parent Frame be one of
+  // the new clones if there is a match. This is meant to minimize any possible
+  // interdependencies between the kinematics of different worlds.
+  for(size_t i=0; i<worldClone->getNumFrames(); ++i)
   {
-    delete (*it);
+    dynamics::Frame* current_parent =
+        worldClone->getFrame(i)->getParentFrame();
+
+    dynamics::SimpleFramePtr parent_candidate =
+        worldClone->getFrame(current_parent->getName());
+
+    if(parent_candidate)
+      worldClone->getFrame(i)->setParentFrame(parent_candidate.get());
```

Comment:
> It's more like, a SimpleFrame isn't necessarily tied to a World at all.

I understand it more clearly now. I exclusion the case a SimpleFrame isn't tied to a World.

lazy copying sounds interesting, and I agree with that the implementation would be somewhat challenging.

> The big question would be whether changing a property in the original should also carry over to the existing clones of the original.

In my view, changing a property in the original shouldn't carry over to the clones, and they should be able to have different property values. It might be able to give a flag whether the property carry over to the clones, but keeping same property values in different world might be physically infeasible under certain circumstances. However, all the entities in the master world will always carry the same properties to all the dependent worlds. One use case I can imagine now is that we put entities for static environment to the master world and run different simulations in multiple worlds sharing the mast world. So whether carrying a property or not will be the differences between lazy copying and master world.